river/drift/adwin_c.pyx

```diff
@@ -0,0 +1,379 @@
+from typing import Deque
+from collections import deque
+
+
+cdef class ADWINC:
+    """ The helper class for ADWIN
+
+    Parameters
+    ----------
+    delta
+        Confidence value.
+
+    """
+    cdef:
+        dict __dict__
+        double delta, total, variance, total_width, width
+        int n_buckets, min_window_len,\
+            mint_min_window_length, tick, n_detections,\
+            clock, max_n_buckets, detect, detect_twice
+
+    MAX_BUCKETS = 5
+
+    def __init__(self, delta=.002):
+        # default values affected by init_bucket()
+        self.delta = delta
+        self.bucket_deque = None
+        self.total = 0.
+        self.variance = 0.
+        self.width = 0.
+        self.n_buckets = 0
+
+        self._init_buckets()
+
+        # other default values
+        self.min_window_len = 10
+
+        self.tick = 0
+        self.total_width = 0
+
+        self.detect = 0
+        self.n_detections = 0
+        self.detect_twice = 0
+        self.clock = 32
+
+        self.max_n_buckets = 0
+        self.mint_min_window_length = 5
+
+    def reset(self):
+        """Reset the change detector.
+        """
+        self.__init__(delta=self.delta)
+
+    def get_delta(self):
+        return self.delta
+
+    def get_n_detections(self):
+        return self.n_detections
+
+    def get_width(self):
+        return self.width
+
+    def get_total(self):
+        return self.total
+
+    def get_variance(self):
+        return self.variance
+
+    @property
+    def variance_in_window(self):
+        return self.variance / self.width
+
+    cdef void _init_buckets(self):
+        """ Initialize the bucket's list and statistics
+
+        Set all statistics to 0 and create a new bucket List.
+
+        """
+        self.bucket_deque: Deque['Bucket'] = deque([Bucket()])
+        self.total = 0.0
+        self.variance = 0.0
+        self.width = 0.0
+        self.n_buckets = 0
+
+    def update(self, value: float):
+        """Update the change detector with a single data point.
+
+        Apart from adding the element value to the window, by inserting it in
+        the correct bucket, it will also update the relevant statistics, in
+        this case the total sum of all values, the window width and the total
+        variance.
+
+        Parameters
+        ----------
+        value
+            Input value
+
+        Returns
+        -------
+        bool
+            If True then a change is detected.
+
+        """
+        return self._update(value)
+
+    cdef bint _update(self, double value):
+
+        cdef double incremental_variance
+
+        self.width += 1
+
+        self._insert_element(value, 0.0)
+        incremental_variance = 0.0
+        if self.width > 1.0:
+            incremental_variance = (
+                    (self.width - 1.0)
+                    * (value - self.total / (self.width - 1.0))
+                    * (value - self.total / (self.width - 1.0))
+                    / self.width
```

Comment:
This one is similar to what we do in `stats.Var`: the current estimation is firstly scaled up by `ddof - 1` and the incremental update is applied. Not so sure about the last denominator.